File: src/playnode.cpp
#include "ros/ros.h"
#include "std_msgs/String.h"
#include <message_filters/subscriber.h>
#include <message_filters/time_synchronizer.h>
#include <message_filters/sync_policies/approximate_time.h>
#include <sensor_msgs/LaserScan.h>
#include "sensor_msgs/Image.h"
#include <sensor_msgs/image_encodings.h>
#include <geometry_msgs/Twist.h>
#include <image_transport/image_transport.h>
#include "cv_bridge/cv_bridge.h"
#include "opencv2/highgui/highgui.hpp"
#include "opencv2/opencv.hpp"

#include <image_transport/image_transport.h>
#include <image_transport/subscriber_filter.h>

// Segment Added by Reza//
// Aims at Re-Exporting in-game Objects
#include <pcl/sample_consensus/mlesac.h>
#include <pcl_ros/point_cloud.h>
#include <pcl/sample_consensus/sac_model_circle3d.h> // Deploys 3D_Circular Feature Models
//#include <SacModel.h>

typedef typedef pcl::PointXYZRGB Point_RGB_3D;
typedef pcl::PointCloud<pcl::PointXYZRGB>::Ptr Cloud_PTr;
typedef pcl::PointCloud<Point_RGB_3D>::ConstPtr Cloud_ConstPTr;



typedef pcl::PointCloud<pcl::PointXYZRGB> PointCloud;
typedef pcl::io::loadPCDFile PCL_Path;
typedef pcl::visualization::PCLVisualizer PCL_Visualizer;

typedef pcl::MaximumLikelihoodSampleConsensus< Point_RGB_3D > MaximumLikelihood;
typedef pcl::SampleConsensusModelCircle3D< Point_RGB_3D> Circular_M3D;
typedef boost::shared_ptr< SampleConsensusModelCircle3D< PointT > > Circular_M3D_Ptr;

class Cloud_Processor
{

private:
    ros::Subscriber kinect_sub;


public:
    PointCloud_Processor ()
    {




public:



        kinect_sub = n->subscribe("robot1/kinect/depth_registered/points", 1,
                                  &PlayNode::kinect_callback, this);
        Cloud_PTr raw_Cloud (new PointCloud);
        Cloud_PTr filtered_Cloud (new PointCloud);

        const Cloud_ConstPTr  circular_ReferenceCloud (& new PointCloud);




        PointCloud filtered_Cloud;




        Circular_M3D cicle3D new (circular_ReferenceCloud, random= false));
        MaximumLikelihood cylindricalLikelyHood= new MaximumLikelihood(& );
        cylindricalLikelyHood::ComputeModel(0);






        processed_type1=Cloud_PTr (new PointCloud);


    }

    sensor_msgs::LaserScan laser_msg;
    bool got_image;
    bool got_laser;

    void image_callback(const sensor_msgs::Image::ConstPtr& msg_img)
    {
        ROS_INFO("Got new image");
        cv_bridge::CvImagePtr cv_ptr = cv_bridge::toCvCopy(msg_img, sensor_msgs::image_encodings::BGR8);
        // Store the latest image received - this variable always has the most up to
        // date image.
        image = cv_ptr->image;

        got_image = true;
    }









}

void laser_callback(const sensor_msgs::LaserScan::ConstPtr& msg_laser)
{
    ROS_INFO("Got new laser scan");
    // Store the laser message we get. The laser_msg variable always holds the
    // latest laser message.
    laser_msg = *msg_laser;
    got_laser = true;
}

void set_velocities(float lin_vel, float ang_vel)
{
    geometry_msgs::Twist msg;
    msg.linear.x = lin_vel;
    msg.angular.z = ang_vel;

    velocity_pub.publish(msg);
}

void show_image()
{
    cv::imshow(IMAGE_WINDOW, image);
    cv::waitKey(10);
}


int main(int argc, char **argv)
{
    // The node name is robot_node
    ros::init(argc, argv, "robot_node");
    // The nodehandle actually starts the node
    ros::NodeHandle n;

    // Publish messages to move around
    velocity_pub = n.advertise<geometry_msgs::Twist>("cmd_vel", 1000);

    ROS_INFO("Waiting for camera image");
    ros::topic::waitForMessage<sensor_msgs::Image>("robot1/front_camera/image_raw");
    ROS_INFO("Waiting for laser scan message");
    ros::topic::waitForMessage<sensor_msgs::LaserScan>("robot1/front_laser/scan");

    // Images need a special subscriber
    image_transport::ImageTransport it(n);
    image_transport::Subscriber image_sub = it.subscribe("robot1/front_camera/image_raw", 1, image_callback);

    // basic subscriber for the laser
    ros::Subscriber laser_sub = n.subscribe("robot1/front_laser/scan", 1, laser_callback);

    cv::namedWindow(IMAGE_WINDOW);

    // 10Hz loop
    ros::Rate r(10);
    while(ros::ok())
    {
        if (got_laser && got_image)
        {
            // Copy the laser and image. These change whenever the callbacks run, and
            // we don't want the data changing in the middle of processing.
            cv::Mat cur_img = image;
            //sensor_msgs::LaserScan cur_laser = laser_msg;



            show_image();
        }


        ros::spinOnce();
        r.sleep();
    }

    return 0;
}

File: scripts/playnode.py
#!/usr/bin/env python
import sys

import rospy
from std_msgs.msg import String
from sensor_msgs.msg import LaserScan, Image
from geometry_msgs.msg import Twist
import message_filters

import cv2
from cv_bridge import CvBridge, CvBridgeError

class PlayNode:

    
    def __init__(self, image_window="Camera Input", message_slop=0.1, synchroniser_queuesize=20):
        """
        :param message_slop: Messages with a header.stamp within message_slop
        seconds of each other will be considered to be synchronisable
        """
        rospy.init_node("robot_node")
        rospy.loginfo("Initialised PlayNode")
        
        self.bridge = CvBridge()
        self.image_window = image_window

        self.velocity_pub = rospy.Publisher("cmd_vel", Twist, queue_size=1000)
        self.image_sub = message_filters.Subscriber("front_camera/image_raw", Image)
        self.laser_sub = message_filters.Subscriber("front_laser/scan", LaserScan)

        self.time_sync = message_filters.ApproximateTimeSynchronizer([self.image_sub, self.laser_sub],
                                                                     synchroniser_queuesize,
                                                                     message_slop)
        self.time_sync.registerCallback(self.perception_cb)

    def perception_cb(self, img_msg, laser_msg):
        rospy.loginfo("Received new image ({}) and scan ({})".format(img_msg.header.stamp, laser_msg.header.stamp))
        
        try:
            image = self.bridge.imgmsg_to_cv2(img_msg, "bgr8")
        except CvBridgeError as e:
            rospy.logerr(e)
            return

        image = cv2.flip(image, -1)
        cv2.imshow(self.image_window, image)
        cv2.waitKey(10)
        
        # DO SOMETHING WITH image AND laser_msg HERE... OR NOT...

    def set_velocities(linear, angular):
        msg = Twist()
        msg.linear.x = linear
        msg.angular.z = angular
        self.velocity_pub.publish(msg)

if __name__ == '__main__':

    play_node = PlayNode()
       
    loop_rate = rospy.Rate(10)

    rospy.loginfo("Starting loop")
    while not rospy.is_shutdown():

        # 10 Hz loop
        
        loop_rate.sleep()

